### ISSUE TITLE
IsolationForest panic: 0 ClassAttributes (1 expected)

### ISSUE THREAD

**Issue by andrewm4894** [sjwhitworth/golearn#260]
Hello,

I'm playing around and trying to get an IsolationForest working for some anomaly detection. I seem to be running into an issue whereby i dont have any ClassAttributes set. Looking at the example [here](https://github.com/sjwhitworth/golearn/blob/master/examples/trees/isolationForest/isolation_forest.go) i don't really see anything specific about setting class attributes or anything and don't see any obvious fields in the example gaussian csv file either. 

It feels like it wants me to set a class attribute but i don't think i should need to for Isolation Forest no?

Here is my code as an example:

```Go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"sync"

	"github.com/sjwhitworth/golearn/base"
	"github.com/sjwhitworth/golearn/trees"
	"gonum.org/v1/gonum/mat"
)

// Create a wait group
var wg sync.WaitGroup

// LagsN defines number of lags to make
var LagsN = 2

// Urls define a list of api calls we want data from
var Urls = [1]string{
	//"https://london.my-netdata.io/api/v1/data?chart=system.cpu&format=json&after=-4",
	"https://london.my-netdata.io/api/v1/data?chart=system.net&format=json&after=-200",
	//"https://london.my-netdata.io/api/v1/data?chart=system.load&format=json&after=-4",
	//"https://london.my-netdata.io/api/v1/data?chart=system.io&format=json&after=-3",
}

type netdataResponse struct {
	Labels []string    `json:"labels"`
	Data   [][]float64 `json:"data"`
}

// Get a gonum matrix from the netdata api with specified nLags
func getX(url string, nLags int, c chan mat.Matrix) {

	// Create data to store response
	var data netdataResponse

	// Need to make sure we tell wait group we done
	defer wg.Done()

	// Get response
	resp, _ := http.Get(url)
	bodyBytes, _ := ioutil.ReadAll(resp.Body)

	// Unmarshal into dataMap
	_ = json.Unmarshal([]byte(bodyBytes), &data)

	// Flatten data into one slice, ignoring the first column which is "time", and adding nLags
	nData := len(data.Data)
	nDims := len(data.Labels) - 1
	nCols := nDims + (nLags * nDims)
	nRows := nData - nLags

	// Make flat slice to put data into
	dataFlat := make([]float64, nCols*nRows)

	// Loop over and add lags to flat data
	i := 0
	for t := range data.Data {
		fmt.Println(data.Data[t])
		if t >= nLags {
			for dim := range data.Data[t] {
				// Ignore time which is the first dim in the response
				if dim > 0 {
					// Add each lag
					for l := 0; l <= nLags; l++ {
						dataFlat[i] = data.Data[t-l][dim]
						i++
					}
				}
			}
		}
	}

	// Create gonum dense matrix from dataFlat
	x := mat.NewDense(nRows, nCols, dataFlat)

	// Send to channel
	c <- x

}

func main() {

	// Create a channel the size of number of api calls we need to make
	dataChannel := make(chan mat.Matrix, len(Urls))

	// Kick off a go routine for each url
	for _, url := range Urls {
		wg.Add(1)
		go getX(url, LagsN, dataChannel)
	}

	// Handle synchronization of channel
	wg.Wait()
	close(dataChannel)

	// Pull each response from channel
	for x := range dataChannel {

		// Print matrix x
		fmt.Printf("X:\n %v\n\n", mat.Formatted(x, mat.Prefix(" "), mat.Excerpt(10)))

		// Create instances
		r, c := x.Dims()
		instances := base.InstancesFromMat64(r, c, mat.DenseCopyOf(x))
		fmt.Println(instances)

		// Create forest
		forest := trees.NewIsolationForest(10, 10, 100)

		// Fit forest
		forest.Fit(instances)

		// Predict on instances to get scores
		preds := forest.Predict(instances)

		// Let's find the average and minimum Anomaly Score for normal data
		var avgScore float64
		var min float64
		min = 1
		for i := 0; i < len(preds); i++ {
			temp := preds[i]
			avgScore += temp
			if temp < min {
				min = temp
			}
		}
		fmt.Println(avgScore / 1000)
		fmt.Println(min)

		fmt.Println("Anomaly Scores are ")
		for i := range preds {
			fmt.Println(preds[i])
		}

	}

}

```

And my output:
```
➜  learn-go git:(main) ✗ go run scripts/netdataGoNum.go
[1.609452054e+09 77.81402 -117.20517]
[1.609452053e+09 95.26177 -182.4566]
[1.609452052e+09 183.7658 -149.93954]
[1.609452051e+09 63.5156 -106.90236]
[1.60945205e+09 105.81358 -268.4757]
[1.609452049e+09 95.26471 -228.5053]
[1.609452048e+09 149.17166 -77.10375]
...
...
[1.609451885e+09 18.02243 -13.352751]
[1.609451884e+09 33.20156 -130.75358]
[1.609451883e+09 127.26549 -312.6093]
[1.609451882e+09 180.4365 -133.4438]
[1.609451881e+09 7.980919 -10.109821]
[1.60945188e+09 12.7004 -67.86969]
[1.609451879e+09 30.57802 -234.2778]
[1.609451878e+09 118.38961 -251.7482]
[1.609451877e+09 188.8146 -84.43673]
[1.609451876e+09 27.79418 -132.28655]
[1.609451875e+09 17.50297 -16.629344]
[1.609451874e+09 46.46548 -52.69503]
[1.609451873e+09 138.70405 -126.53562]
[1.609451872e+09 227.7109 -186.4979]
[1.609451871e+09 55.26046 -48.6947]
[1.60945187e+09 36.8918 -33.45698]
[1.609451869e+09 33.983 -42.40188]
[1.609451868e+09 135.65547 -76.88999]
[1.609451867e+09 183.6133 -30.55653]
[1.609451866e+09 17.85453 -20.09791]
[1.609451865e+09 38.74968 -44.52087]
[1.609451864e+09 62.9204 -74.47051]
[1.609451863e+09 102.00025 -32.32786]
[1.609451862e+09 183.9357 -84.19254]
[1.609451861e+09 33.92577 -147.15892]
[1.60945186e+09 32.15584 -54.7018]
[1.609451859e+09 13.1178 -9.887853]
[1.609451858e+09 92.97254 -22.72811]
[1.609451857e+09 194.8131 -40.32558]
[1.609451856e+09 17.69448 -34.85213]
[1.609451855e+09 12.857935 -136.00853]
X:
 Dims(198, 6)
 ⎡  183.7658    95.26177    77.81402  -149.93954   -182.4566  -117.20517⎤
 ⎢   63.5156    183.7658    95.26177  -106.90236  -149.93954   -182.4566⎥
 ⎢ 105.81358     63.5156    183.7658   -268.4757  -106.90236  -149.93954⎥
 ⎢  95.26471   105.81358     63.5156   -228.5053   -268.4757  -106.90236⎥
 ⎢ 149.17166    95.26471   105.81358   -77.10375   -228.5053   -268.4757⎥
 ⎢  208.5272   149.17166    95.26471  -110.52662   -77.10375   -228.5053⎥
 ⎢  62.14063    208.5272   149.17166   -181.5678  -110.52662   -77.10375⎥
 ⎢  99.34486    62.14063    208.5272  -102.04994   -181.5678  -110.52662⎥
 ⎢  64.42528    99.34486    62.14063   -120.6335  -102.04994   -181.5678⎥
 ⎢  95.16334    64.42528    99.34486    -59.2533   -120.6335  -102.04994⎥
  .
  .
  .
 ⎢   62.9204    38.74968    17.85453   -74.47051   -44.52087   -20.09791⎥
 ⎢ 102.00025     62.9204    38.74968   -32.32786   -74.47051   -44.52087⎥
 ⎢  183.9357   102.00025     62.9204   -84.19254   -32.32786   -74.47051⎥
 ⎢  33.92577    183.9357   102.00025  -147.15892   -84.19254   -32.32786⎥
 ⎢  32.15584    33.92577    183.9357    -54.7018  -147.15892   -84.19254⎥
 ⎢   13.1178    32.15584    33.92577   -9.887853    -54.7018  -147.15892⎥
 ⎢  92.97254     13.1178    32.15584   -22.72811   -9.887853    -54.7018⎥
 ⎢  194.8131    92.97254     13.1178   -40.32558   -22.72811   -9.887853⎥
 ⎢  17.69448    194.8131    92.97254   -34.85213   -40.32558   -22.72811⎥
 ⎣ 12.857935    17.69448    194.8131  -136.00853   -34.85213   -40.32558⎦

Instances with 198 row(s) 6 attribute(s)
Attributes:
        FloatAttribute(0)
        FloatAttribute(1)
        FloatAttribute(2)
        FloatAttribute(3)
        FloatAttribute(4)
        FloatAttribute(5)

Data:
        183.77 95.26 77.81 -149.94 -182.46 -117.21
        63.52 183.77 95.26 -106.90 -149.94 -182.46
        105.81 63.52 183.77 -268.48 -106.90 -149.94
        95.26 105.81 63.52 -228.51 -268.48 -106.90
        149.17 95.26 105.81 -77.10 -228.51 -268.48
        208.53 149.17 95.26 -110.53 -77.10 -228.51
        62.14 208.53 149.17 -181.57 -110.53 -77.10
        99.34 62.14 208.53 -102.05 -181.57 -110.53
        64.43 99.34 62.14 -120.63 -102.05 -181.57
        95.16 64.43 99.34 -59.25 -120.63 -102.05
        173.32 95.16 64.43 -54.71 -59.25 -120.63
        7.46 173.32 95.16 -7.78 -54.71 -59.25
        29.81 7.46 173.32 -37.05 -7.78 -54.71
        55.71 29.81 7.46 -83.94 -37.05 -7.78
        107.86 55.71 29.81 -110.74 -83.94 -37.05
        192.31 107.86 55.71 -138.21 -110.74 -83.94
        23.45 192.31 107.86 -22.33 -138.21 -110.74
        15.75 23.45 192.31 -69.69 -22.33 -138.21
        30.16 15.75 23.45 -170.31 -69.69 -22.33
        109.99 30.16 15.75 -85.85 -170.31 -69.69
        183.24 109.99 30.16 -34.19 -85.85 -170.31
        21.14 183.24 109.99 -33.23 -34.19 -85.85
        63.41 21.14 183.24 -80.36 -33.23 -34.19
        95.98 63.41 21.14 -69.06 -80.36 -33.23
        99.39 95.98 63.41 -25.71 -69.06 -80.36
        174.95 99.39 95.98 -75.84 -25.71 -69.06
        30.05 174.95 99.39 -134.12 -75.84 -25.71
        10.29 30.05 174.95 -12.47 -134.12 -75.84
        14.40 10.29 30.05 -31.22 -12.47 -134.12
        97.43 14.40 10.29 -44.69 -31.22 -12.47
        ...
168 row(s) undisplayed
panic: 0 ClassAttributes (1 expected)

goroutine 1 [running]:
github.com/sjwhitworth/golearn/trees.preprocessData(0x7b8b40, 0xc0004f8330, 0xc0000920c0, 0xc0001a4c00, 0x5d8)
        /home/andre/go/src/github.com/sjwhitworth/golearn/trees/isolation.go:224 +0x1f8
github.com/sjwhitworth/golearn/trees.(*IsolationForest).Fit(0xc0000fdf48, 0x7b8b40, 0xc0004f8330)
        /home/andre/go/src/github.com/sjwhitworth/golearn/trees/isolation.go:139 +0x50
main.main()
        /home/andre/repos/learn-go/scripts/netdataGoNum.go:115 +0x64e
exit status 2
```

I'm new to Go and Golearn so pretty sure i might be doing something silly somewhere but can't really see anything obvious. 

I might try just add a column of all '0's and maybe try add that as the class and see if that fixes it which i guess could be ok if just means i'm assuming all my training data to be 'normal' which is reasonable. 

Anyway just wanted to see if i could get some advice or pointers as to if this is some sort of issue or just me doing something a little wrong.

Cheers,
Andy

**Comment by andrewm4894** [comment]
Update on this one - i got a sort of hacky workaround working i think:

```Go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"sync"

	"github.com/sjwhitworth/golearn/base"
	"github.com/sjwhitworth/golearn/trees"
	"gonum.org/v1/gonum/mat"
)

// Create a wait group
var wg sync.WaitGroup

// LagsN defines number of lags to make
var LagsN = 2

// Urls define a list of api calls we want data from
var Urls = [1]string{
	//"https://london.my-netdata.io/api/v1/data?chart=system.cpu&format=json&after=-4",
	"https://london.my-netdata.io/api/v1/data?chart=system.net&format=json&after=-10",
	//"https://london.my-netdata.io/api/v1/data?chart=system.load&format=json&after=-4",
	//"https://london.my-netdata.io/api/v1/data?chart=system.io&format=json&after=-3",
}

type netdataResponse struct {
	Labels []string    `json:"labels"`
	Data   [][]float64 `json:"data"`
}

// Get a gonum matrix from the netdata api with specified nLags
func getX(url string, nLags int, c chan mat.Dense) {

	// Create data to store response
	var data netdataResponse

	// Need to make sure we tell wait group we done
	defer wg.Done()

	// Get response
	resp, _ := http.Get(url)
	bodyBytes, _ := ioutil.ReadAll(resp.Body)

	// Unmarshal into dataMap
	_ = json.Unmarshal([]byte(bodyBytes), &data)

	// Flatten data into one slice, ignoring the first column which is "time", and adding nLags
	nData := len(data.Data)
	nDims := len(data.Labels) - 1
	nCols := nDims + (nLags * nDims)
	nRows := nData - nLags

	// Make flat slice to put data into
	dataFlat := make([]float64, nCols*nRows)

	// Loop over and add lags to flat data
	i := 0
	for t := range data.Data {
		fmt.Println(data.Data[t])
		if t >= nLags {
			for dim := range data.Data[t] {
				// Ignore time which is the first dim in the response
				if dim > 0 {
					// Add each lag
					for l := 0; l <= nLags; l++ {
						dataFlat[i] = data.Data[t-l][dim]
						i++
					}
				}
			}
		}
	}

	// Create gonum dense matrix from dataFlat
	x := mat.NewDense(nRows, nCols, dataFlat)

	// Create vector of zeros to use as a dummy class attribute for golearn
	var xFinal mat.Dense
	zeros := make([]float64, nRows)
	z := mat.NewVecDense(nRows, zeros)
	xFinal.Augment(z, x)
	//fmt.Printf("xFinal:\n %v", xFinal)

	// Send to channel
	c <- xFinal

}

func main() {

	// Create a channel the size of number of api calls we need to make
	dataChannel := make(chan mat.Dense, len(Urls))

	// Kick off a go routine for each url
	for _, url := range Urls {
		wg.Add(1)
		go getX(url, LagsN, dataChannel)
	}

	// Handle synchronization of channel
	wg.Wait()
	close(dataChannel)

	// Pull each response from channel
	for x := range dataChannel {

		// Create instances
		r, c := x.Dims()
		instances := base.NewDenseCopy(base.InstancesFromMat64(r, c, &x))

		// Set a class attribute
		attrArray := instances.AllAttributes()
		instances.AddClassAttribute(attrArray[0])

		// Look at instances
		fmt.Println(instances)

		// Create forest
		forest := trees.NewIsolationForest(10, 10, 100)

		// Fit forest
		forest.Fit(instances)

		// Predict on instances to get scores
		preds := forest.Predict(instances)

		// Let's find the average and minimum Anomaly Score for normal data
		var avgScore float64
		var min float64
		min = 1
		for i := 0; i < len(preds); i++ {
			temp := preds[i]
			avgScore += temp
			if temp < min {
				min = temp
			}
		}
		fmt.Println(avgScore / 1000)
		fmt.Println(min)

		fmt.Println("Anomaly Scores are ")
		for i := range preds {
			fmt.Println(preds[i])
		}

	}

}

```

Which gives:
```
➜  learn-go git:(main) ✗ go run scripts/netdataGolearn.go
[1.609537808e+09 99.69887 -23.02496]
[1.609537807e+09 183.4741 -33.48168]
[1.609537806e+09 17.46232 -66.95385]
[1.609537805e+09 18.37029 -180.6296]
[1.609537804e+09 17.37755 -123.59229]
[1.609537803e+09 105.55665 -37.93892]
[1.609537802e+09 192.9129 -131.94132]
[1.609537801e+09 65.04145 -197.7705]
[1.6095378e+09 53.87894 -80.21452]
[1.609537799e+09 62.1585 -133.25699]
Instances with 8 row(s) 7 attribute(s)
Attributes:
*       FloatAttribute(0)
        FloatAttribute(1)
        FloatAttribute(2)
        FloatAttribute(3)
        FloatAttribute(4)
        FloatAttribute(5)
        FloatAttribute(6)

Data:
        0.00 17.46 183.47 99.70 -66.95 -33.48 -23.02
        0.00 18.37 17.46 183.47 -180.63 -66.95 -33.48
        0.00 17.38 18.37 17.46 -123.59 -180.63 -66.95
        0.00 105.56 17.38 18.37 -37.94 -123.59 -180.63
        0.00 192.91 105.56 17.38 -131.94 -37.94 -123.59
        0.00 65.04 192.91 105.56 -197.77 -131.94 -37.94
        0.00 53.88 65.04 192.91 -80.21 -197.77 -131.94
        0.00 62.16 53.88 65.04 -133.26 -80.21 -197.77
All rows displayed
0.004341934126608612
0.5187841492103988
Anomaly Scores are
0.5407589735670113
0.5502426571317391
0.5187841492103988
0.5410175528845719
0.5718919437242114
0.5436871966572925
0.5496077489659361
0.5259439044674505
```

Just wondering is there a simpler or cleaner way as not quite sure i fully understand some of the types i'm using like mat.Dense and mat.Matrix and feels like i am flipping back and forth a bit that might add some overhead maybe. 

Or would i be safe to just not need to put in the col of 0's and just use the first feature vector as a class attribute and be safe to assume will have no impact on the isolation forest? Had a look at the source code but i don't really understand Go enough yet to be sure if that a safe assumption or not.

**Comment by Sentimentron** [comment]
Creating a column of zeros should work OK, I agree that the API is a bit suboptimal (at the time when I wrote this core API in ~2014 or so, there wasn't really a DataFrame-style library available for Golang, maybe something to look at?).

**Comment by Yushgoel** [comment]
@andrewm4894 I actually wrote the isolation forest code in the library, and I chose to require a class attribute because of the following:
When reading a csv into a FixedDataGrid, golearn automatically chooses one of the columns as a class attribute. Since isolation forest is unsupervised and we don't want to lose on any data, I chose to include the class attribute as part of the training data.

Creating a vector of 0s is fine, but it can impact performance (you are adding a new feature where every data point is the same so there is no anomaly with regards to this single feature). Better than that, just choose one of the normal features and make it the class attribute. Rest of the code should work just fine.
